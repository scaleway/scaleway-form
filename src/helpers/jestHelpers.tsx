import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as scwuiTheme } from '@scaleway/ui'
import { RenderResult, render } from '@testing-library/react'
import React, { FC, ReactElement, ReactNode } from 'react'
import Form from '../components/Form'
import mockErrors from '../mocks/mockErrors'

interface WrapperProps {
  theme?: typeof scwuiTheme
}

const Wrapper: FC<WrapperProps> = ({ theme = scwuiTheme, children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: {
    transform: (node: RenderResult) => Promise<void> | void
  },
) =>
  shouldMatchEmotionSnapshot(
    <Form errors={mockErrors}>{() => children}</Form>,
    options,
  )

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={scwuiTheme}>
        <Form errors={mockErrors}>{children}</Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
