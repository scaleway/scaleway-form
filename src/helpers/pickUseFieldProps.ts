import { AnyObject } from 'final-form'
import { UseFieldConfig } from 'react-final-form'
import { BaseFieldProps } from '../types'

const pickUseFieldProps = <T = AnyObject, K = AnyObject>({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
}: BaseFieldProps<T, K>): UseFieldConfig<T, K> => ({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
})

export default pickUseFieldProps
