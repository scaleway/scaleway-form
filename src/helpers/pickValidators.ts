import type { ValidatorObject, ValidatorProps } from '../types'
import validators from '../validators'
import type { ValidatorFn } from '../validators/types'

export const pickValidators = <InputValue = unknown>(args: ValidatorProps) =>
  Object.entries(args)
    .map(([key, value]) =>
      value !== undefined
        ? (validators as Record<string, ValidatorFn>)[
            key as keyof typeof validators
          ]?.(value)
        : undefined,
    )
    .filter(validator => !!validator) as ValidatorObject<InputValue>[]
