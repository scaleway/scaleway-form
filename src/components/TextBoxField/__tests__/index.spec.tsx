import { act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TextBoxField } from '..'
import { shouldMatchEmotionSnapshotFormWrapper } from '../../../../.jest/helpers'
import { mockErrors } from '../../../mocks'

describe('TextBoxField', () => {
  test('should render correctly', () =>
    shouldMatchEmotionSnapshotFormWrapper(<TextBoxField name="test" />))

  test('should render correctly generated', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TextBoxField name="test" generated />,
    ))

  test('should render correctly random', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TextBoxField name="test" random="random" />,
    ))
  test('should render correctly notice', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TextBoxField name="test" notice="notice" />,
    ))

  test('should render correctly required', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TextBoxField name="test" required />,
    ))
  test('should render correctly id', () =>
    shouldMatchEmotionSnapshotFormWrapper(<TextBoxField name="test" id="id" />))

  test('should render correctly disabled', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TextBoxField name="test" disabled />,
      {
        transform: node => {
          const input = node.getByRole('textbox')
          expect(input).toBeDisabled()
        },
      },
    ))

  test('should render correctly with minLength', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TextBoxField name="test" minLength={13} />,
      {
        transform: async node => {
          await act(async () => {
            const input = node.getByRole('textbox')
            await userEvent.type(input, 'test')
            input.blur()
          })
          expect(
            node.getByText(
              typeof mockErrors.MIN_LENGTH === 'function'
                ? mockErrors.MIN_LENGTH({
                    allValues: {},
                    label: 'test',
                    minLength: 13,
                    name: 'test',
                    value: 'test',
                  })
                : mockErrors.MIN_LENGTH,
            ),
          ).toBeVisible()
        },
      },
    ))
})
