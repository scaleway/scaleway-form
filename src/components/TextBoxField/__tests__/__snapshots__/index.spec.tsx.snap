// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBoxField should render correctly 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-16jruyc {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
}

.cache-16jruyc::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:hover,
.cache-16jruyc:focus {
  border-color: #4f0599;
}

.cache-16jruyc:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-16jruyc::-webkit-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::-moz-placeholder {
  opacity: 1;
}

.cache-16jruyc:-ms-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::placeholder {
  opacity: 1;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-16jruyc e18bmnpq1"
          name="test"
          type="text"
          value=""
        />
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly disabled 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-4sr5xp {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  cursor: default;
  pointer-events: none;
  background-color: #f6f6f8;
  border-color: #d4dae7;
  color: #c4c9d6;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
}

.cache-4sr5xp::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-4sr5xp::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-4sr5xp:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-4sr5xp::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-4sr5xp:hover,
.cache-4sr5xp:focus {
  border-color: #4f0599;
}

.cache-4sr5xp:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-4sr5xp::-webkit-input-placeholder {
  opacity: 1;
}

.cache-4sr5xp::-moz-placeholder {
  opacity: 1;
}

.cache-4sr5xp:-ms-input-placeholder {
  opacity: 1;
}

.cache-4sr5xp::placeholder {
  opacity: 1;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-4sr5xp e18bmnpq1"
          disabled=""
          name="test"
          type="text"
          value=""
        />
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly generated 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-16jruyc {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
}

.cache-16jruyc::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:hover,
.cache-16jruyc:focus {
  border-color: #4f0599;
}

.cache-16jruyc:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-16jruyc::-webkit-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::-moz-placeholder {
  opacity: 1;
}

.cache-16jruyc:-ms-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::placeholder {
  opacity: 1;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-16jruyc e18bmnpq1"
          name="test"
          type="text"
          value=""
        />
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly id 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-16jruyc {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
}

.cache-16jruyc::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:hover,
.cache-16jruyc:focus {
  border-color: #4f0599;
}

.cache-16jruyc:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-16jruyc::-webkit-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::-moz-placeholder {
  opacity: 1;
}

.cache-16jruyc:-ms-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::placeholder {
  opacity: 1;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-16jruyc e18bmnpq1"
          id="id"
          name="test"
          type="text"
          value=""
        />
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly notice 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-16jruyc {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
}

.cache-16jruyc::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-16jruyc:hover,
.cache-16jruyc:focus {
  border-color: #4f0599;
}

.cache-16jruyc:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-16jruyc::-webkit-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::-moz-placeholder {
  opacity: 1;
}

.cache-16jruyc:-ms-input-placeholder {
  opacity: 1;
}

.cache-16jruyc::placeholder {
  opacity: 1;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

.cache-1wy3rxk {
  color: #b2b6c3;
  font-size: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  margin-top: 8px;
}

.cache-qq4de9-sizeStyles {
  vertical-align: middle;
  fill: currentColor;
  height: 20px;
  width: 20px;
  min-width: 20px;
  min-height: 20px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-16jruyc e18bmnpq1"
          name="test"
          type="text"
          value=""
        />
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
      <div
        class="e18bmnpq3 cache-1wy3rxk e1qg14600"
      >
        <svg
          class="cache-qq4de9-sizeStyles etwatq50"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        <div
          class="cache-1xdhyk6 e1kf57p60"
        >
          <p>
            <span>
              notice
            </span>
          </p>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly random 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-z48xia {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
  padding-right: 32px;
}

.cache-z48xia::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-z48xia::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-z48xia:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-z48xia::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-z48xia:hover,
.cache-z48xia:focus {
  border-color: #4f0599;
}

.cache-z48xia:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-z48xia::-webkit-input-placeholder {
  opacity: 1;
}

.cache-z48xia::-moz-placeholder {
  opacity: 1;
}

.cache-z48xia:-ms-input-placeholder {
  opacity: 1;
}

.cache-z48xia::placeholder {
  opacity: 1;
}

.cache-zmd8bf-StyledRightElement {
  pointer-events: none;
  position: absolute;
  right: 0;
  bottom: 0;
  top: 0;
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: -webkit-transform 150ms,color 150ms;
  transition: transform 150ms,color 150ms;
  color: #b2b6c3;
  pointer-events: auto;
}

.cache-zmd8bf-StyledRightElement:hover,
.cache-zmd8bf-StyledRightElement:focus-within {
  color: #4a4f62;
}

.cache-zmd8bf-StyledRightElement>button {
  box-shadow: none!important;
}

.cache-19k31i2-StyledButton {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border-width: 0;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  vertical-align: middle;
  white-space: nowrap;
  font-weight: 500;
  -webkit-transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  font-size: 16px;
  line-height: 32px;
  font-weight: 500;
  padding: 8px 16px;
  background-color: transparent;
  color: #4a4f62;
  width: 32px;
  height: 32px;
  padding: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.cache-19k31i2-StyledButton:hover,
.cache-19k31i2-StyledButton:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.cache-19k31i2-StyledButton>svg {
  pointer-events: none;
}

.cache-brwenz {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  pointer-events: none;
}

.cache-kv8vgz-sizeStyles {
  vertical-align: middle;
  fill: currentColor;
  height: 1em;
  width: 1em;
  min-width: 1em;
  min-height: 1em;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-z48xia e18bmnpq1"
          name="test"
          type="text"
          value=""
        />
        <div
          class="cache-zmd8bf-StyledRightElement e18bmnpq7"
        >
          <button
            aria-disabled="false"
            class="cache-19k31i2-StyledButton em6gco80"
            title="Randomize"
            type="button"
          >
            <div
              class="cache-brwenz em6gco81"
            >
              <svg
                class="cache-kv8vgz-sizeStyles etwatq50"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.5,5.6L5,7L6.4,4.5L5,2L7.5,3.4L10,2L8.6,4.5L10,7L7.5,5.6M19.5,15.4L22,14L20.6,16.5L22,19L19.5,17.6L17,19L18.4,16.5L17,14L19.5,15.4M22,2L20.6,4.5L22,7L19.5,5.6L17,7L18.4,4.5L17,2L19.5,3.4L22,2M13.34,12.78L15.78,10.34L13.66,8.22L11.22,10.66L13.34,12.78M14.37,7.29L16.71,9.63C17.1,10 17.1,10.65 16.71,11.04L5.04,22.71C4.65,23.1 4,23.1 3.63,22.71L1.29,20.37C0.9,20 0.9,19.35 1.29,18.96L12.96,7.29C13.35,6.9 14,6.9 14.37,7.29Z"
                />
              </svg>
            </div>
          </button>
        </div>
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly required 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-t4dkq1 {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
  padding-right: 22px;
}

.cache-t4dkq1::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-t4dkq1::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-t4dkq1:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-t4dkq1::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-t4dkq1:hover,
.cache-t4dkq1:focus {
  border-color: #4f0599;
}

.cache-t4dkq1:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-t4dkq1::-webkit-input-placeholder {
  opacity: 1;
}

.cache-t4dkq1::-moz-placeholder {
  opacity: 1;
}

.cache-t4dkq1:-ms-input-placeholder {
  opacity: 1;
}

.cache-t4dkq1::placeholder {
  opacity: 1;
}

.cache-h0isfn-StyledRightElement {
  pointer-events: none;
  position: absolute;
  right: 0;
  bottom: 0;
  top: 0;
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: -webkit-transform 150ms,color 150ms;
  transition: transform 150ms,color 150ms;
  color: #b2b6c3;
}

.cache-h0isfn-StyledRightElement:hover,
.cache-h0isfn-StyledRightElement:focus-within {
  color: #4a4f62;
}

.cache-7wqnxk-sizeStyles {
  vertical-align: middle;
  fill: #a6102d;
  height: 10px;
  width: 10px;
  min-width: 10px;
  min-height: 10px;
}

.cache-9vn1ol {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-t4dkq1 e18bmnpq1"
          name="test"
          type="text"
          value=""
        />
        <div
          class="cache-h0isfn-StyledRightElement e18bmnpq7"
        >
          <svg
            class="cache-7wqnxk-sizeStyles etwatq50"
            viewBox="0 0 24 24"
          >
            <path
              d="M10,2H14L13.21,9.91L19.66,5.27L21.66,8.73L14.42,12L21.66,15.27L19.66,18.73L13.21,14.09L14,22H10L10.79,14.09L4.34,18.73L2.34,15.27L9.58,12L2.34,8.73L4.34,5.27L10.79,9.91L10,2Z"
            />
          </svg>
        </div>
      </div>
      <div
        class="cache-9vn1ol e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`TextBoxField should render correctly with minLength 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-1l46zoq {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
  border-color: #dd3252;
}

.cache-1l46zoq::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq:hover,
.cache-1l46zoq:focus {
  border-color: #4f0599;
}

.cache-1l46zoq:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-1l46zoq::-webkit-input-placeholder {
  opacity: 1;
}

.cache-1l46zoq::-moz-placeholder {
  opacity: 1;
}

.cache-1l46zoq:-ms-input-placeholder {
  opacity: 1;
}

.cache-1l46zoq::placeholder {
  opacity: 1;
}

.cache-1l46zoq:hover,
.cache-1l46zoq:focus {
  border-color: #a6102d;
}

.cache-1l46zoq:focus {
  box-shadow: 0px 0px 0px 3px #DD325240;
  border-color: #a6102d;
}

.cache-msb198 {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  -webkit-transition: max-height 300ms ease-in,opacity 300ms ease-in;
  transition: max-height 300ms ease-in,opacity 300ms ease-in;
  max-height: 56px;
  opacity: 1;
  overflow: visible;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-1l46zoq e18bmnpq1"
          name="test"
          type="text"
          value="test"
        />
      </div>
      <div
        class="cache-msb198 e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        >
          This field should have a length greater than 13
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
